### PR TITLE
feat: filter products

### PR DESCRIPTION
**Pull request by samflab**
This PR contains: 
1. Filter Context
2. Filter Reducer (Hook)
3. Filter Methods:
  - Sort by Price
      - High to Low
      - Low to High
   - Sort by Rating
      - High to Low
      - Low to High
   - Filtering through various Categories
   - Price Range Slider
   - Reset Filter
5.  A toggle filter option (works in mobile view)

`Note: I have not added a message when no products are rendered on any selected filter combination.`

### CODE REVIEW

File: src/components/ProductList/Filter.js [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+          </div>
+        </div>
+      ) : (
+        <> </>
+      )}
+    </>
+  );
+}
+
+export const categoryData = [
```
Comment (rutvikpumak) [inline src/components/ProductList/Filter.js]: move this data to somewhere else

File: src/components/ProductList/ProductList.js [hunk 2]
```diff
@@ -15,33 +18,42 @@ export const ProductList = () => {
   useEffect(() => {
     getProducts();
   }, []);
+
+  const { state } = useFilter();
+  const filterData = filterMethod(state, product);
+
+  console.log("data", filterData)
```
Comment (rutvikpumak) [inline src/components/ProductList/ProductList.js]: remove console

File: src/components/ProductList/ProductList.js [hunk 3]
```diff
@@ -15,33 +18,42 @@ export const ProductList = () => {
...
-      }}
-    >
-      {product.map((i) => {
-        const discount = Math.ceil(
-          ((i.originalPrice - i.price) / i.originalPrice) * 100
-        );
+    <>
+      <Filter/>
+      <div
+        style={{
```
Comment (rutvikpumak) [inline src/components/ProductList/ProductList.js]: remove inline styling

**Review by rutvikpumak — COMMENTED** [review]
It looks confusing, rating and price high, to low and low to high

File: src/components/ProductList/Filter.js [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+                  payload: "low-to-high-price",
+                })
+              }
+              value="low-high-price"
+              name="price-sort"
+              id="low-high-price"
+            />
+            <label htmlFor="price-sort" className="checkbox">
+              Price: low to high
+            </label>
```
Comment (narharikale) [inline src/components/ProductList/Filter.js]: why not wrap the `input` element inside the `label`? , is there any specific reason

File: src/components/ProductList/Filter.js [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+                Clear
+              </button>
+            </h3>
+            <input
+              type="radio"
+              checked={priceSort === "low-to-high-price"}
+              onChange={() =>
+                dispatch({
+                  type: "PRICE_SORT",
+                  payload: "low-to-high-price",
```
Comment (mohammadkhizerkhan) [inline src/components/ProductList/Filter.js]: give payload as "LOW_TO_HIGH_PRICE"
kushank told us to do this

File: src/context/filter-context.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+import { filterReducer } from "../hooks/filterReducer";
+
+const FilterContext = createContext();
+
+const FilterProvider = ({children}) => {
+    const [state, dispatch] = useReducer(filterReducer, {
+        priceSort: "",
+        category: [],
+        ratingSort: "",
+        range: 2000
```
Comment (mohammadkhizerkhan) [inline src/context/filter-context.js:11]: why are you initializing range from 2000??
is your products price starts from 2000???

File: src/components/ProductList/ProductList.js [hunk 7]
```diff
@@ -15,33 +18,40 @@ export const ProductList = () => {
...
+          flexWrap: "wrap",
+          gap: "1rem",
+          padding: "2rem",
+          justifyContent: "center",
+        }}
+      >
+        {filterData.map((i) => {
+          const discount = Math.ceil(
+            ((i.originalPrice - i.price) / i.originalPrice) * 100
+          );
```
Comment (narharikale) [inline src/components/ProductList/ProductList.js]: can we change name of `i`  it would be more readable and clear

**Review by mohammadkhizerkhan — CHANGES_REQUESTED** [review]

**Review by narharikale — APPROVED** [review]
code is structure good ✔

File: src/components/ProductList/Filter.js [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+  return (
+    <>
+      <button onClick={() => setShowFilter(() => !showFilter)} class="filter">
+        <i class="fas fa-sort"></i> Filters
+      </button>
+      {showFilter ? (
+        <div className="filter-div">
+          <div>
+            <h3 className="filter-heading">
+              <u>Price</u>
```
Comment (shraddha15-creator) [inline src/components/ProductList/Filter.js]: We can write *sort by* instead of Price

**Review by mohammadkhizerkhan — APPROVED** [review]

File: src/context/filter-context.js (hunk 6 above)
Comment (samflab) [reply src/context/filter-context.js:11]: no, that's the max price. without applying any filters all the products under 2000 will be shown.

**Pull request merged**